kafka-streams-framework/src/main/java/org/hypertrace/core/kafkastreams/framework/AsyncTransformer.java

```diff
@@ -0,0 +1,96 @@
+package org.hypertrace.core.kafkastreams.framework;
+
+import com.google.common.util.concurrent.RateLimiter;
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+import java.time.Duration;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ArrayBlockingQueue;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+import java.util.function.Supplier;
+import lombok.SneakyThrows;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.kafka.streams.KeyValue;
+import org.apache.kafka.streams.kstream.Transformer;
+import org.apache.kafka.streams.processor.ProcessorContext;
+
+@Slf4j
+@SuppressWarnings("UnstableApiUsage")
+public abstract class AsyncTransformer<K, V, KOUT, VOUT>
+    implements Transformer<K, V, KeyValue<KOUT, VOUT>> {
+  private final Executor executor;
+  private final BlockingQueue<CompletableFuture<List<KeyValue<KOUT, VOUT>>>> pendingFutures;
+  private final RateLimiter rateLimiter;
+  private ProcessorContext context;
+
+  // TODO: configurable executor - supplier pattern. This enables to use a common thread-pool for
+  // all stream tasks
+  public AsyncTransformer(int concurrency, int maxBatchSize, Duration flushInterval) {
+    ThreadFactory threadFactory =
+        new ThreadFactoryBuilder()
+            .setNameFormat("async-transformer-pool-%d")
+            .setDaemon(true)
+            .build();
+    this.executor = Executors.newFixedThreadPool(concurrency, threadFactory);
```

Comment:
What is the range of configuring `concurrency` we're expecting? (initial estimate, ofc tuning might be required)

CompletableFuture offloads all work to the executor pool from the main stream thread (Q: even the chained computation runs on offloaded thread right?) and those individual threads will blocked while loading file / doing GRPC call via a blocking stub. 

Technically speaking we're not doing async programming but just increasing through put by parallelising blocking calls. Which is fine, but there is a noticeable cost to PlatformThreads due to kernel involvement hence doubt if we can keep the number high say in close to kafka batch size. So just trying to understand what the reasonable values here are for `concurrency`.

However, with Java 21 LTS releasing this September we can simulate close to async programming (dep on GRPC impl support, File IO will be taken care of by JDK) with support of VirtualThreads (aka efficient Green Threads managed by runtime similar to Go). They have much less overhead and outperform PlatformThreads with IO bound computation, (don't expect them to be worse off for CPU bound as well if their executor is utilised properly). 

So thinking a bit towards future, can adopt this transformer in straight forward fashion whenever they get released if we migrate to latest Java release (at least for kafka stream apps). We currently have our base image at Java 11. Wondering why we haven't upgraded to latter LTS releases like 17, did we encounter issues or was there no need for it @aaron-steinfeld ? (just curious: jvm & compiler mostly seem to improve in performance over releases, so shouldn't we be upgrading every lts release)

Reply:
> Wondering why we haven't upgraded to latter LTS releases like 17, did we encounter issues or was there no need for it @aaron-steinfeld ? (just curious: jvm & compiler mostly seem to improve in performance over releases, so shouldn't we be upgrading every lts release)

Basically hasn't been a priority. There's better support for the older LTS versions before they started shipping new releases every 6 months. I was going to upgrade us to 17 just never got around to it, then when I had time I figured I might as well wait for 21  :) We don't want to do a blind upgrade, and we want to enable developers on new features, so we need to be thoughtful on cadence. I think 21 would be good time to move forward.

In terms of this feature, I don't think we're talking a huge number of threads here - single digits unless I misunderstood. Does the thread overhead still seem concerning to you?

Reply:
> In terms of this feature, I don't think we're talking a huge number of threads here - single digits unless I misunderstood. Does the thread overhead still seem concerning to you?

Nope, that shouldn't be of any concern.

Reply:
> In terms of this feature, I don't think we're talking a huge number of threads here - single digits unless I misunderstood. Does the thread overhead still seem concerning to you?

Nope, that shouldn't be of any concern.